tests/unit/nupic/algorithms/knn_classifier_test.py

```diff
@@ -0,0 +1,147 @@
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import unittest
+
+import numpy as np
+
+from nupic.algorithms.KNNClassifier import KNNClassifier
+
+
+
+class KNNClassifierTest(unittest.TestCase):
+
+
+  def testOverlapDistanceMethod_Standard(self):
+    """Tests standard learning case for raw overlap"""
+    params = {"distanceMethod": "rawOverlap"}
+    classifier = KNNClassifier(**params)
+
+    dimensionality = 40
+    a = np.array([1, 3, 7, 11, 13, 17, 19, 23, 29], dtype=np.int32)
+    b = np.array([2, 4, 8, 12, 14, 18, 20, 28, 30], dtype=np.int32)
+
+    numPatterns = classifier.learn(a, 0, isSparse=dimensionality)
+    self.assertEquals(numPatterns, 1)
+
+    numPatterns = classifier.learn(b, 1, isSparse=dimensionality)
+    self.assertEquals(numPatterns, 2)
+
+    denseA = np.zeros(dimensionality)
+    denseA[a] = 1.0
+    cat, inferenceResult, dist, categoryDist = classifier.infer(denseA)
+    self.assertEquals(cat, 0)
+
+    denseB = np.zeros(dimensionality)
+    denseB[b] = 1.0
+    cat, inferenceResult, dist, categoryDist = classifier.infer(denseB)
+    self.assertEquals(cat, 1)
+
+
+  def testOverlapDistanceMethod_BadSparsity(self):
+    """Sparsity (input dimensionality) less than input array"""
+    params = {"distanceMethod": "rawOverlap"}
+    classifier = KNNClassifier(**params)
+
+    dimensionality = 40
+    a = np.array([1, 3, 7, 11, 13, 17, 19, 23, 29], dtype=np.int32)
+
+    # Learn with incorrect dimensionality, less than some bits (23, 29)
+    with self.assertRaises(RuntimeError):
+      classifier.learn(a, 0, isSparse=20)
+
+
+  def testOverlapDistanceMethod_InconsistentSparsity(self):
+    """Inconsistent sparsity (input dimensionality)"""
+    params = {"distanceMethod": "rawOverlap"}
+    classifier = KNNClassifier(**params)
+
+    dimensionality = 40
+    a = np.array([1, 3, 7, 11, 13, 17, 19, 23, 29], dtype=np.int32)
+
+    # Learn with incorrect dimensionality, greater than largest ON bit, but
+    # inconsistent when inferring
+    numPatterns = classifier.learn(a, 0, isSparse=31)
+    self.assertEquals(numPatterns, 1)
+
+    denseA = np.zeros(dimensionality)
+    denseA[a] = 1.0
+
+    cat, inferenceResult, dist, categoryDist = classifier.infer(denseA)
+    self.assertEquals(cat, 0)
```

Comment:
It's about inconsistentSparsity or inconsistentDimensionality
The idea here was to learn on a vector with isSparse=31 and then infer with dimensionality of 40. isSparse is an unfortunate name because it's actually referring to dimensionality....